lib/common/patchset.c

```diff
@@ -711,42 +706,53 @@ void
 pcmk__xml_log_patchset(uint8_t log_level, const xmlNode *patchset)
 {
     int format = 1;
+    pcmk__output_t *out = NULL;
 
-    if (log_level == LOG_NEVER) {
-        return;
+    static struct qb_log_callsite *patchset_cs = NULL;
+
+    switch (log_level) {
+        case LOG_NEVER:
+            return;
+        case LOG_STDOUT:
+            CRM_CHECK(pcmk__text_output_new(&out, NULL) == pcmk_rc_ok, return);
+            break;
+        default:
+            if (patchset_cs == NULL) {
+                patchset_cs = qb_log_callsite_get(__func__, __FILE__,
+                                                  "xml-patchset", log_level,
+                                                  __LINE__, 0);
```

Comment:
Probably could use `crm_trace_nonlog`, yeah. Still haven't gone down the full rabbit hole of the difference that makes.

Hmm. A few thoughts, and correct me if I'm wrong.
* `crm_is_callsite_active()`'s helper functions (e.g., `crm_log_filter_source()`) specifically take `crm_trace_nonlog` into consideration, so some thought may have already gone into handling it properly within that call chain.
* The filtering in `crm_is_callsite_active()` is what sets `cs->targets` after the initial `qb_log_callsite_get()`. The log level might differ between calls to `pcmk__xml_log_patchset()`, so I think we'd want to be able to refilter and adjust whether `cs->targets == 0` or not. Change of either log level or tags triggers a refilter, which is desirable I think.
* If we haven't changed the log level or the tags, then we won't refilter at all. For example, if we initialize the callsite with `qb_log_callsite_get(..., LOG_TRACE, ..., crm_trace_nonlog)` and then we call `crm_is_callsite_active(cs, LOG_TRACE, crm_trace_nonlog)`, then `crm_is_callsite_active()` is a quick call that doesn't do any filtering -- it returns true if `(cs != NULL) && (cs->targets != 0)`, or false otherwise.
* If we do call `crm_is_callsite_active()` in places where we created the callsite with `crm_trace_nonlog` as the tag (in `qb_log_callsite_get()`), we should probably pass `crm_trace_nonlog` instead of `0` to `crm_is_callsite_active()` too. Right now we're passing 0 as the tag to `crm_is_callsite_active()` for callsites that were initialized with tag `crm_trace_nonlog`.

Reply:
The commit on top is just playing around with `crm_trace_nonlog` under the *supposition* that `crm_is_callsite_active()` behaves correctly when the `tags` argument is `crm_trace_nonlog`. Just a quick demo and there may be bugs.

Easy to revert it and move all the `crm_trace_nonlog` cases to use `(cs != NULL) && (cs->targets != 0)`.

Reply:
> Probably could use `crm_trace_nonlog`, yeah. Still haven't gone down the full rabbit hole of the difference that makes.
> 
> Hmm. A few thoughts, and correct me if I'm wrong.
> 
>     * `crm_is_callsite_active()`'s helper functions (e.g., `crm_log_filter_source()`) specifically take `crm_trace_nonlog` into consideration, so some thought may have already gone into handling it properly within that call chain.

It looks like it, but it takes a circuitous path there, so better to avoid if possible.

>       * This gets questionable with regard to `trace_files`, `trace_fns`, and `trace_fmts`... `crm_log_filter_source()` doesn't care whether `crm_trace_nonlog` is the tag if we find a match in those cases.

Those are irrelevant if not logging, so it's not a problem, just not necessary either.

> 
>     * The filtering in `crm_is_callsite_active()` is what sets `cs->targets` after the initial `qb_log_callsite_get()`. The log level might differ between calls to `pcmk__xml_log_patchset()`, so I think we'd want to be able to refilter and adjust whether `cs->targets == 0` or not. Change of either log level or tags triggers a refilter, which is desirable I think.

`qb_log_callsite_get()` sets cs->priority and cs->tags, and it looks like it sets cs->targets via `_log_filter_apply_to_cs()` the first time the callsite is created.

So, I think it makes sense to use qb_log_callsite_get() and crm_is_callsite_active(), just using crm_trace_nonlog in both for non-logging tests. If the log level changes after the initial callsite creation, it'll refilter.

>     * If we haven't changed the log level or the tags, then we won't refilter at all. For example, if we initialize the callsite with `qb_log_callsite_get(..., LOG_TRACE, ..., crm_trace_nonlog)` and then we call `crm_is_callsite_active(cs, LOG_TRACE, crm_trace_nonlog)`, then `crm_is_callsite_active()` is a quick call that doesn't do any filtering -- it returns true if `(cs != NULL) && (cs->targets != 0)`, or false otherwise.

Yes, after a closer look, I do think the function is worth using everywhere for consistency
 
>     * If we do call `crm_is_callsite_active()` in places where we created the callsite with `crm_trace_nonlog` as the tag (in `qb_log_callsite_get()`), we should probably pass `crm_trace_nonlog` instead of `0` to `crm_is_callsite_active()` too. Right now we're passing 0 as the tag to `crm_is_callsite_active()` for callsites that were initialized with tag `crm_trace_nonlog`.

Agreed

Reply:
> > This gets questionable with regard to `trace_files`, `trace_fns`, and `trace_fmts`... `crm_log_filter_source()` doesn't care whether `crm_trace_nonlog` is the tag if we find a match in those cases.

> Those are irrelevant if not logging, so it's not a problem, just not necessary either.

My concern was that regardless of whether we're logging or not, the trace variables are set. If they match some attribute of the callsite, then we'll set a bit of `cs->targets`. That results in `crm_is_callsite_active()` returning true.

But that was a silly concern; I must have been spacing out. The whole point is to return true if tracing is enabled. If we match against a trace variable, then tracing is enabled. So I think we're good here.

> `qb_log_callsite_get()` sets `cs->priority` and `cs->tags`, and it looks like it sets `cs->targets` via `_log_filter_apply_to_cs()` the first time the callsite is created.

Right, my wording wasn't great. I meant that after that initialization step, the filtering triggered by `crm_is_callsite_active()` is what updates `cs->targets` -- so we want that in case the priority changes.

---

I'll take a closer look at that last commit, write a better commit message, and update.